### ISSUE TITLE
Hilbert transforms are extremely slow on odd-length signals

### ISSUE THREAD

**Issue by npyoung** [scipy/scipy#6324]
Starting with input array lengths in the tens of thousands, `scipy.signal.hilbert` is orders of magnitude slower when operating on odd-length arrays compared to even. The problem is especially bad when array lengths are far from powers of two. To be clear: it's not that the Hilbert transform is much slower when lengths are not near powers of 2, it's only slower on odd-length arrays that are not near powers of 2. Here's an example on my desktop:

``` python
In [1]: import numpy as np
In [2]: import scipy  
In [3]: np.__version__, scipy.__version__  
Out[3]: ('1.10.4', '0.17.1') 

In [4]: from scipy.signal import hilbert 
In [5]: signal = np.random.randn(50000) 
In [6]: %timeit hilbert(signal) 
100 loops, best of 3: 1.95 ms per loop 

In [7]: %timeit hilbert(signal[:-1])
1 loops, best of 3: 6.24 s per loop
```

That's **3200 times slower** on an array with one _fewer_ element. I don't believe this behavior is an inherent feature of computing the Hilbert Transform.

**Comment by WarrenWeckesser** [comment]
The Hilbert transform code spends most of its time computing a fast Fourier transform and its inverse, and these are the functions that have the funky time dependence on the signal length.  Here, for example, is the time to compute just `ifft(fft(signal))`:

```
In [18]: from scipy.fftpack import fft, ifft

In [19]: signal = np.random.randn(50000)

In [20]: %timeit ifft(fft(signal))
1000 loops, best of 3: 1.41 ms per loop

In [21]: %timeit ifft(fft(signal[:-1]))
1 loop, best of 3: 5.68 s per loop
```

**Comment by npyoung** [comment]
Should `hilbert` be doing some kind of zero padding or is `fft` doing this under the hood?

**Comment by haphaeu** [comment]
You can force zero padding using the `N` argument, if that is what you want:

```
import numpy as np
import scipy  

from scipy.signal import hilbert 
signal = np.random.randn(50000) 

%timeit hilbert(signal) 
%timeit hilbert(signal[:-1])
%timeit hilbert(signal[:-1], N=50000)

#  output:
#  100 loops, best of 3: 2.78 ms per loop
#  1 loops, best of 3: 6.07 s per loop
#  100 loops, best of 3: 2.79 ms per loop
```

**Comment by npyoung** [comment]
That works for me. I thought it would do some zero padding automatically.

It doesn't crop the result back to the original number of samples, though. If you give it a signal of length N_samples and run `result = hilbert(signal, N=N_fft)`, the correct thing to ultimately return is `result[:N_samples]` i. e. tail crop, correct?

**Comment by haphaeu** [comment]
I'd rather leave that part to the user. Note you can zero pad, or also truncate. In the latter the function should not invent values to bring the result back to size `N_samples`.

**Comment by pv** [comment]
If the results are the same, it would make sense to zero pad 
automatically --- and cut results to same size, unlike what the `N` 
parameter does. This is what eg. fftconvolve does.

**Comment by npyoung** [comment]
I agree with @pv - it's a good idea to emulate the behavior of fftconvolve.

**Comment by endolith** [comment]
> That's 3200 times slower on an array with one fewer element.

49999 is a prime number, so yes, FFT will be extremely slow.  

> If the results are the same, it would make sense to zero pad automatically --- and cut results to same size, unlike what the `N` parameter does. This is what eg. fftconvolve does.

Yes, it's just fft → throw away half → ifft, right?  So padding first and truncating after should produce the same result?

**Comment by Eric89GXL** [comment]
I'm pretty sure it should produce the same result, yeah. It would be nice
to have some auto mode that used a next fast length and truncated back.

**Comment by jmaralc** [comment]
Hi, 
@endolith was right. I think padding with zeros and chunk the result is the best option. At least the best I found. The FFT is only efficient if you use a signal whose length is power of 2. Hence padding with zeros unitl the next power of 2 should solve this problem. 

I just code a simple function to do that and return only the chunk corresponding to the original signal:

def Hilbert(signal):
    padding = np.zeros(int(2 ** np.ceil(np.log2(len(signal)))) - len(signal))
    tohilbert = np.hstack((signal, padding))
    
    result = sgnl.hilbert(tohilbert)
    
    result = result[0:len(signal)]

    return result

Thanks for the discussions guys I was like three hours trying to understand why I did not have the same result as in Matlab. :)

**Comment by endolith** [comment]
@jmaralc 
> The FFT is only efficient if you use a signal whose length is power of 2. 

Actually you can use http://scipy.github.io/devdocs/generated/scipy.fftpack.next_fast_len.html to find the best size

**Comment by jmaralc** [comment]
Thanks @endolith! I did not know about it.

**Comment by rsandler00** [comment]
Agree with most on the thread. This should be automatically done ...

**Comment by josephernest** [comment]
This solves it:

    hilbert3 = lambda x: signal.hilbert(x, fftpack.next_fast_len(len(x)))[:len(x)]

Don't call it `hilbert2`, because it already exists ;)

**Comment by peterbell10** [comment]
Fixed by #10361
